tables/TaQL/ExprNodeSetOpt.h

```diff
@@ -0,0 +1,358 @@
+
+  // <prerequisite>
+  //# Classes you should understand before using this one.
+  //   <li> TableExprNodeSet
+  // </prerequisite>
+
+  // <synopsis>
+  // This class is an optimized representation of an constant integer array set
+  // with a large range of values used by the IN operator.
+  // If applicable, TableExprLogicNode instantiates an object of this class
+  // for sets with a value range of more than 16384.
+  // <br>The representation is a std::unordered_map containing the array values.
+  // <br>Note that a std::unordered_map is used instead of std::set because its
+  // hashing mechanism makes it faster.
+  // </synopsis>
+
+  class TableExprNodeSetOptBase : public TableExprNodeRep
+  {
+  public:
+    explicit TableExprNodeSetOptBase (const TableExprNodeRep& orig);
+    virtual Bool contains (const TableExprId& id, Int64 value) override;
+    virtual Bool contains (const TableExprId& id, Double value) override;
+    virtual Bool contains (const TableExprId& id, String value) override;
+    virtual MArray<Bool> contains (const TableExprId& id,
+                                   const MArray<Int64>& value) override;
+    virtual MArray<Bool> contains (const TableExprId& id,
+                                   const MArray<Double>& value) override;
+    virtual MArray<Bool> contains (const TableExprId& id,
+                                   const MArray<String>& value) override;
+    // Tell which key matches a value. -1 = no match.
+    // The default implementations throw an exception.
+    virtual Int64 find (Int64 value) const;
+    virtual Int64 find (Double value) const;
+    virtual Int64 find (String value) const;
+  };
+
+
+  // <summary>
+  // An optimized representation of a discrete selection set.
+  // </summary>
+
+  // <use visibility=local>
+
+  // <reviewed reviewer="UNKNOWN" date="before2004/08/25" tests="">
+  // </reviewed>
+
+  // <prerequisite>
+  //# Classes you should understand before using this one.
+  //   <li> TableExprNodeSet
+  // </prerequisite>
+
+  // <synopsis>
+  // This templated class is an optimized representation of an constant
+  // integer or string array set used by the IN operator.
+  // If applicable, TableExprLogicNode instantiates an object of this class.
+  // <br>The representation is a std::unordered_map containing the array values
+  // and the index in the array.
+  // <br>Note that a std::unordered_map is used instead of std::map because its
+  // hashing mechanism makes it faster.
+  // </synopsis>
+
+  template <typename T>
+  class TableExprNodeSetOptUSet: public TableExprNodeSetOptBase
+  {
+  public:
+    // Construct an empty set.
+    TableExprNodeSetOptUSet (const TableExprNodeRep& orig, const Array<T>&);
+
+    // Show the node.
+    virtual void show (ostream& os, uInt indent) const override;
+
+    // Where does a value occur in the set? -1 is no match.
+    virtual Int64 find (T value) const override;
+
+  private:
+    std::unordered_map<T,Int64> itsMap;
```

Comment:
It's a map now.
For TableExprNodeSet a set would suffice because it only needs to know if the value exists. But I'm also going to use it for joins, which need to know which value matches.
I'll update the comments in the header file.